### PR TITLE
TypeNode#annotated_types

### PR DESCRIPTION
**Pull request by Blacksmoke16**
Returns an array of all types (struct/class/module) with the given annotation.  Addresses some of #7274.  

```crystal
annotation Foo
end

@[Foo(val: 1)]
class Bar
end

@[Foo(val: 2)]
class Baz
end

class Blah
end

{% for a in Foo.annotated_types %}
  pp {{a.annotation(Foo)[:val]}}
{% end %}

# => 1
# => 2
```

A future PR could add `#annotated_methods`/`#annotated_variables`, but `#annotated_types` is the more useful one IMO.

/cc @vladfaust

### CODE REVIEW

**Comment by vladfaust** [comment]
Small note – you don't actually check for *types* in specs, only for their size. I'm also concerned of how does it work with #7326...

**Comment by Blacksmoke16** [comment]
In the case of #7326 you actually want to keep duplicates since the values on each could be different.  However, `annotated_types` you only want the list of types that are annotated, you wouldn't want `Klass` to show up twice just because you added two annotations to it.

Even if those two annotations had separate values, you would still be able to use `#annotations` in the for loop to get access to them.

**Comment by j8r** [comment]
This can be confused with classic type annotations. The choice over Attributes was decided in https://github.com/crystal-lang/crystal/pull/6063

**Comment by asterite** [comment]
Just a note that things like this push us further and further away from incremental/modular compilation (at this point we should probably drop that idea altogether and accept the existing nature of Crystal).

**Comment by j8r** [comment]
@asterite if each library/shard is independent from each other, won't it be possible to compile each one and link them to the main program?

**Comment by Blacksmoke16** [comment]
@asterite @straight-shoota anything else need to be done?

**Comment by RX14** [comment]
@asterite incremental macros just means careful tracking of exactly which AST nodes a macro "sees" and being able to verify they're unchanged. I think that can be automated fairly well. We should remove the main blocker for incremental/modular compilation (type restrictions only being restrictions) *before* blocking other features over that though.

I think this is good to go.

File: src/compiler/crystal/types.cr [hunk 1]
```diff
@@ -2685,6 +2685,8 @@ module Crystal
   end
 
   class AnnotationType < NamedType
+    property class_types : Set(Type) = Set(Type).new
```
Comment (straight-shoota) [inline src/compiler/crystal/types.cr]: Is `class_types` a good name for this? Why not call it `annotated_type` as well?

Comment (Blacksmoke16) [reply src/compiler/crystal/types.cr]: Its a set of the classes that have a given annotation, so makes sense yea.

**Comment by bcardiff** [comment]
I would like to have the ability to query types/ivars/methods given an annotation type. 

I think that the incremental is doable (at some cost).

But what I would like to have a discussion/design before merging this is regarding subtyping. If Bar < Foo, are annotations inherited or not? Would that behavior be better specified in the annotation declaration itself (like in C# IIRC) or to define that upon querying (adding params to the proposed API)

File: src/compiler/crystal/types.cr (hunk 1 above)
Comment (straight-shoota) [reply src/compiler/crystal/types.cr]: => `annotated_types`

**Comment by Blacksmoke16** [comment]
I think it would be better if they were not, and just added to each type explicitly; however I could be convinced otherwise.  IMO, This would make it less likely for unintended things to happen.  Being more verbose but allowing for more control seems to be a fair trade to me.

**Comment by Blacksmoke16** [comment]
Renamed that property @straight-shoota, CI failure seems unrelated?

**Comment by Blacksmoke16** [comment]
I don't suppose this will make it into `0.29.0`?

**Comment by bcardiff** [comment]
@Blacksmoke16 I am interested in the feature but I haven't invested time yet to have proper design discussions. From my past experience, I am still inclined to mimic C# interface. But we should also consider how other language features impact: an annotation that is applied in a module that is included in a class does contribute to the annotations?

**Comment by Blacksmoke16** [comment]
I'd like to get the discussion around this started again.  I'm getting close to a point where this is kind of a blocker to some Athena stuff I'm working on.  #8133 will be a good temporary solution, but having both this and that would be ideal.

**Comment by watzon** [comment]
Any movement on this? This really is probably the ideal way for annotations to work, since they won't be hamstrung by needing to loop over annotations inside of a class. I really need this working for a project as well.

**Comment by kalinon** [comment]
> Any movement on this? This really is probably the ideal way for annotations to work, since they won't be hamstrung by needing to loop over annotations inside of a class. I really need this working for a project as well.

Yes +1 for this feature

**Comment by asterite** [comment]
I think this will be needed for the new logging stuff, otherwise how do you find all registered logs?

(I wonder why there's no registry and instead the proposal is using annotations, but it's a separate issue).

**Comment by watzon** [comment]
It does seem necessary for a number of things, and it's so close to approval

**Comment by bcardiff** [comment]
@asterite it's not needed. The following works.

```crystal
  def self.build_backend(type : String, config : Backend::Config)
    {% begin %}
      case type
      {% for t in Log::Backend.all_subclasses %}
        {% if reg = t.annotation(Log::RegisterBackend) %}
          when {{reg[0]}}
            {{t}}.new(config)
        {% end %}
      {% end %}
      else
        raise "There is no Log::Backend with @[Log::RegisterBackend(\"#{type}\")] annotation"
      end
    {% end %}
  end
```

It is not a registry as DB drivers because that might lead to a situation where the backend is not registered on time to initialize the `Log` constants.

Back to coding ... 👋

**Comment by asterite** [comment]
So the annotation is only for the name, but you still need to inherit `Log::Backend`. Interesting!

Why not rely on a class method like `backend_name`? Why use annotations which are not type-safe?

**Comment by asterite** [comment]
Also a warning: if a generic type inherits `Log::RegisterBackend` then it will stop working because you won't be able to use that uninstantiated generic type. Maybe there's no use case for generic types, though...

**Comment by bcardiff** [comment]
> Why not rely on a class method like backend_name? Why use annotations which are not type-safe?

I liked the registry. The case is generated with literals. It can be changed of course, it is not the more important aspect :-). But the annotation based registry is doable.

**Comment by Blacksmoke16** [comment]
> Also a warning: if a generic type inherits `Log::RegisterBackend` then it will stop working because you won't be able to use that uninstantiated generic type

This is what #8458 would help with, as it could print `Foo(T)` as `Foo`.

> @asterite it's not needed. The following works.

It works yes, but IMO it seems less than ideal to have to have a parent type/module just to do this; hence this PR.  Annotating a type should be enough, not for the `Log` use case specifically but are cases where it would be nice.

**Comment by straight-shoota** [comment]
@Blacksmoke16 Totally agree. But I don't think anyone made the argument that this PR wasn't useful because there are simple alternatives. It's just not strictly necessary for the `Log::Backend` example. 👍

**Comment by Blacksmoke16** [comment]
As a workaround for now, I realized you could just do `Object.all_subclasses.select &.annotation(MyAnn)`

**Comment by Blacksmoke16** [comment]
Closing you can just use the workaround as defined in https://github.com/crystal-lang/crystal/pull/7648#issuecomment-598167731.

**Pull request closed without merging**